### ISSUE TITLE
Gameday + CV

### ISSUE THREAD

**Issue by phil-lopreiato** [the-blue-alliance/the-blue-alliance#2023]
We'd love to integrate @andrewda's library [frc-livescore](https://github.com/andrewda/frc-livescore) into TBA so we can process webcasts to get current event status and score. 

## Potential Applications

 - Better BlueZone (we know exactly when a match starts)
 - ESPN-like "ticker" along the bottom of TBA showing all the current matches
 - Live scores on team pages/GD Lite elements
 - Persist time-series score data so we can nerd out over more analytics
 - Auto-import offseason data from webcast

## Basic System Design
 - In a GCE container, have frc-livescore (or a program that calls it) running. It takes a webcast url, extracts data at a regular interval (1 second?), and logs it to a file.
  - Wrapper program could use [Livestreamer](http://docs.livestreamer.io/) to pull frames from the webcast at an interval and feed them into frc-livescore, and then parse the returned data. 
 - We have a simple python process also running in the container that tails the file and pushes updates to firebase + persisted storage. This could also be combined with the wrapper program.
 - TBA frontend can consume firebase updates and show cool stuff

## Potential Changes to frc-livescore
 - Update for 2018 game (duh)
 - Allow extracting detailed score breakdown items (this year rotor count + fuel count + touchpad count)
 - Implement extracting team numbers
 - Allow matching the post-match screen to pull the final results.

## Open Questions
 - How many webcasts can we parse at once on a single GCE instance. In addition, what's the right type of instance to use (all the fun pricing trade-offs). 
 - Where do we persist this? Cloud SQL? Streaming inserts to BigQuery? We definitely don't want to persist this in Firebase...
 - How do we start/stop various containers? If we have one event per container (simplest), then we can use our existing "is this webcast live" detection to manage the container lifecycle.

**Comment by gregmarra** [comment]
If this nets out to be expensive due to bandwidth or CPU, we could
theoretically also release the container to people, and allow people to
register for trusted API keys to write through TBA to firebase. This would
let people volunteer their desktop computers' compute :)

**Comment by andrewda** [comment]
@gregmarra I definitely like the idea of crowd-sourcing this, especially if we can make something easy to setup on any computer.

**Comment by andrewda** [comment]
Though according to some talk in Slack, I don't think bandwidth will be an expense issue. @fangeugene roughly calculated $1 per event, assuming standard-1 for 8 hours/day and 2 days/event.

**Comment by gregmarra** [comment]
I would definitely prefer to self host this for simplicity.

**Comment by phil-lopreiato** [comment]
> I would definitely prefer to self host this for simplicity.

Agreed (as long as it's not cost prohibitive to do so). But if worse comes to worse, we could build it so people could sponsor an event's live data by running a container on GCE for us.

I think the interesting first steps here are:
 - How hard would it be to make the changes to frc-livescore I outlined above (@andrewda?)
 - What's the best place to persist timeseries score data?
   - Bigquery + [streaming inserts](https://cloud.google.com/bigquery/streaming-data-into-bigquery). This seems straightforward enough, except that the options for partitioning tables are somewhat limited.
   - Cloud Bigtable. This is Google's hosted nosql (I think it's hbase?) database thing. The docs say [it's a good fit for timeseries data](https://cloud.google.com/bigtable/docs/schema-design-time-series).
 - CloudSQL. We can make a postgress/mySQL relational database.
  - Regular old datastore. 

I'd love @fangeugene's (and any other database experts out there) input on some of the tradeoffs between the above methods. I'm leaning towards bigtable, although I know from experience that hbase schemas need to be deisgned very carefully.

**Comment by JacisNonsense** [comment]
What's the predicted size of the database for a single event (in terms of rows, and what those rows contain)?

I think SQL databases are definitely a good way to go (although I am ~~maybe~~ definitely biased). In terms of language and platform flexibility they're really unparalleled.

I've been having some good experiences with Postgres, although if the datasize is small enough SQLite may make it easier to transfer data in bulk out of the container since it's all in a file on disk (and also has minimal setup)

**Comment by fangeugene** [comment]
Per match, ~150 seconds of whatever score breakdowns are available through the scoring overlay for each alliance.

In 2017: total score, (fractional) fuel points, # rotors, # climbs.
In 2016: total score, tower strength, 5x defense status (0, 1, 2 crossings). 
In 2015: total score only

So rough napkin math: one event = 100 matches * 150 seconds * 2 alliances * ~5 "stats" = ~150,000 "integers" = ~4 MB assuming 32 bit ints and not counting keys, indexes, whatever.

My original thought for this was to put everything in a json blob in the Match object in the regular datastore, but maybe other options are more fitting. I'm not sure.

**Comment by phil-lopreiato** [comment]
Yeah, this won't be all that big of a database. One thing to point out is that we won't be running anything locally on a stream processing worker, rather run them in Google's Cloud and use some python library to make updates.

I agree that putting all this into the Match object in the datastore would be simplest, but we would lose the ability to perform interesting queries and make updating it tricky - the worker process would have to get the datastore object, mutate the json, and write it. It would be nice if we could simply fire off a row insert for each frame we process.

Our use case likely doesn't require enough scale to make dealing with hbase worth it, so we can stick with a solution that people know more about. We can use [Cloud SQL](https://cloud.google.com/sql/), which is Google's fully managed mySQL/Postgress solution.

Regarding queries, we can still talk to a Cloud SQL database via GAE (and cache results in `CachedQueryResult`) and expose it in the API.

**Comment by andrewda** [comment]
Progress for frc-livescore thus far. PR's welcome!

- [x] Implement extracting team numbers
- [ ] Allow extracting detailed score breakdown items (this year rotor count + fuel count + touchpad count)
- [ ] Allow matching the post-match screen to pull the final results.

**Comment by tweirtx** [comment]
Saw this as I was combing through issues for Eugene. A concern for crowdhosting was exposing of a write API key, but this can be alleviated by using something like pyinstaller on an admin's machine to generate a binary.

**Comment by bdaroz** [comment]
True, but the API Key would be embedded in the binary and extractable,

**Comment by tweirtx** [comment]
Hmm

**Comment by Ewpratten** [comment]
What about event specific keys?
